nexus/src/app/sagas/snapshot_create.rs

```diff
@@ -0,0 +1,765 @@
+use crate::external_api::params;
+use crate::saga_interface::SagaContext;
+use crate::{authn, authz, db};
+use anyhow::anyhow;
+use crucible_agent_client::{types::RegionId, Client as CrucibleAgentClient};
+use lazy_static::lazy_static;
+use omicron_common::api::external::Error;
+use serde::Deserialize;
+use serde::Serialize;
+use sled_agent_client::types::{
+    DiskSnapshotRequestBody, InstanceIssueDiskSnapshotRequestBody,
+    VolumeConstructionRequest,
+};
+use slog::warn;
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use steno::new_action_noop_undo;
+use steno::ActionContext;
+use steno::ActionError;
+use steno::ActionFunc;
+use steno::SagaTemplate;
+use steno::SagaTemplateBuilder;
+use steno::SagaType;
+use uuid::Uuid;
+
+pub const SAGA_NAME: &'static str = "snapshot-create";
+
+lazy_static! {
+    pub static ref SAGA_TEMPLATE: Arc<SagaTemplate<SagaSnapshotCreate>> =
+        Arc::new(saga_snapshot_create());
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct Params {
+    pub serialized_authn: authn::saga::Serialized,
+    pub silo_id: Uuid,
+    pub organization_id: Uuid,
+    pub project_id: Uuid,
+    pub create_params: params::SnapshotCreate,
+}
+
+#[derive(Debug)]
+pub struct SagaSnapshotCreate;
+impl SagaType for SagaSnapshotCreate {
+    type SagaParamsType = Arc<Params>;
+    type ExecContextType = Arc<SagaContext>;
+}
+
+/// A no-op action used because if saga nodes fail their undo action isn't run!
+async fn ssc_noop(
+    _sagactx: ActionContext<SagaSnapshotCreate>,
+) -> Result<(), ActionError> {
+    Ok(())
+}
+
+fn saga_snapshot_create() -> SagaTemplate<SagaSnapshotCreate> {
+    let mut template_builder = SagaTemplateBuilder::new();
+
+    // Generate IDs
+    template_builder.append(
+        "snapshot_id",
+        "GenerateSnapshotId",
+        new_action_noop_undo(saga_generate_uuid),
+    );
+
+    template_builder.append(
+        "volume_id",
+        "GenerateVolumeId",
+        new_action_noop_undo(saga_generate_uuid),
+    );
+
+    // Create the Snapshot DB object
+    template_builder.append(
+        "created_snapshot",
+        "CreateSnapshotRecord",
+        ActionFunc::new_action(
+            ssc_create_snapshot_record,
+            ssc_create_snapshot_record_undo,
+        ),
+    );
+
+    // The following two saga actions iterate over the datasets and regions for
+    // a disk and make requests for each tuple, and this violates the saga's
+    // mental model where actions should do one thing at a time and be
+    // idempotent. If only a few of the requests succeed, the saga will leave
+    // things in a partial state because the undo function of a node is not run
+    // when the action fails.
+    //
+    // Use a noop action and an undo, followed by an action + no undo, to work
+    // around this:
+    //
+    // - [noop, undo function]
+    // - [action function, noop]
+    //
+    // With this, if the action function fails, the undo function will run.
```

Comment:
I'm confused about the need for this. I'm reading the body of `ssc_send_snapshot_request`, and it looks like (although it does a bunch of lookups) it issues ONE of:

- `instance_issue_disk_snapshot_request` to a single sled agent, or
- `issue_disk_snapshot_request` to a single sled agent

These operations will either "succeed or fail", right? Isn't it the responsibility of the Sled Agent to ensure they either fully succeed or fail and have no side-effects?

Reply:
Either way, https://github.com/oxidecomputer/omicron/issues/1466 was a similar problem. Hopefully this gets easier with some of the steno changes in the pipeline.

Reply:
You're right, now only the `ssc_start_running_snapshot` / `ssc_start_running_snapshot_undo` does the for loop. Fixed in 77f9929a